Document environment variable naming conventions for resources (#498)
microsoft/aspire.dev: src/frontend/src/content/docs/fundamentals/environment-variables.mdx

Reviewer: Add an import for the LearnMore component and use it here to cross-link to related docs.

seoTitle: Aspire environment variable naming conventions guide
description: Learn how Aspire generates environment variable names for connection strings, endpoints, service discovery, and resource properties.
---

import { Aside } from '@astrojs/starlight/components';

When you use `WithReference` to connect resources in the AppHost, Aspire automatically injects environment variables into the consuming resource. This process is called _configuration injection_.

These environment variable names follow specific conventions based on the referenced resource's name and type. Understanding the conventions is especially useful for applications that read environment variables directly instead of relying on typed Aspire client integrations.

## Naming conventions

Aspire generates environment variables in different formats depending on the type of resource being referenced. The following sections describe each category.

### Connection strings

When you reference a resource that exposes a connection string (such as a database, cache, or messaging resource), Aspire generates an environment variable using the `ConnectionStrings__` prefix:

```txt
ConnectionStrings__{resource-name}
```

> The resource name is used **as-is** (preserving the original casing and hyphens). For example:

Reviewer 2: NB: this causes issues in Azure App Service because the service automatically rewrites the ENV as dashes are not supported on Linux.
For this reason `aspire deploy` will fail with an explanation when trying to deploy such a connection string name. There is a flag to ignore this failure though and go through.

```

The `frontend` resource receives:

| Environment variable | Example value            |
| -------------------- | ------------------------ |
| `MY_API_HTTP`        | `http://localhost:5000`  |
| `MY_API_HTTPS`       | `https://localhost:5001` |

The suffix comes from the endpoint name, not necessarily its URI scheme. For example, a named endpoint called `admin` on `my-api` produces `MY_API_ADMIN`. For resource endpoints, Aspire retains the endpoint suffix even when the resource exposes only one endpoint.

### Service discovery variables

Aspire also generates service discovery variables for .NET service resolution. These use the format:

```txt
services__{resource-name}__{endpoint-key}__{index}
```

The `services` prefix and double-underscore (`__`) separators are fixed, but the resource name preserves the casing used in the AppHost. For example, `AddProject<Projects.Api>("MyApi")` with an HTTP endpoint produces `services__MyApi__http__0`. The endpoint key is the scheme for endpoints named `http` or `https`; otherwise, Aspire uses the endpoint name. Applications that don't use .NET service discovery can read the [endpoint URL variables](#endpoint-urls) instead.

> ### Resource properties

Reviewer 2: Maybe I missed a conversation. In the other parts of the docs (where they are listed) and in code these are called "Connection properties".

```

### Python

```python title="Python — main.py"
import os

# Connection strings
cache_conn = os.getenv("ConnectionStrings__my-cache")

# Endpoint URLs
api_url = os.getenv("MY_API_HTTP")

# Resource properties
db_host = os.getenv("MY_CLICKHOUSE_HOST")
```

### JavaScript / TypeScript

```javascript title="JavaScript — app.js"
// Connection strings (use bracket notation for names with hyphens)
const cacheConn = process.env['ConnectionStrings__my-cache'];

// Endpoint URLs
const apiUrl = process.env.MY_API_HTTP;

// Resource properties
const dbHost = process.env.MY_CLICKHOUSE_HOST;
```

<Aside type="caution">

In JavaScript, `process.env` is an object. Property names containing hyphens require bracket notation: `process.env["ConnectionStrings__my-cache"]`.

</Aside>

## Custom environment variables

If you need different variable names, use `WithEnvironment` to set custom environment variables:

```csharp title="C# — AppHost.cs"
var builder = DistributedApplication.CreateBuilder(args);

var db = builder.AddPostgres("postgres").AddDatabase("my-db");

var api = builder.AddPythonApp("api", "../api", "main.py")
    .WithReference(db)

>     .WithEnvironment("DB_HOST", db.Resource.Parent.PrimaryEndpoint.Property(EndpointProperty.Host))
>     .WithEnvironment("DB_PORT", db.Resource.Parent.PrimaryEndpoint.Property(EndpointProperty.Port));

Reviewer 2: Most connection properties are available as direct properties. This makes it easier to write (even in polyglot):

```suggestion
    .WithEnvironment("DB_HOST", db.Resource.Parent.Host)
    .WithEnvironment("DB_PORT", db.Resource.Parent.Port);
```

---
title: Environment variables
description: Learn how Aspire generates environment variables from resource names and how to access them in your applications.
---

import { Aside, Tabs, TabItem } from '@astrojs/starlight/components';

> When you use `WithReference` to connect resources in the AppHost, Aspire automatically injects environment variables into the consuming resource. The names of these environment variables follow specific conventions based on the resource name and the type of reference. Understanding these conventions is essential, especially for non-.NET applications (Python, Node.js, Go, and others) where you must read environment variables directly rather than using typed Aspire client integrations.

Reviewer: ```suggestion
When you use `WithReference` to connect resources in the AppHost, Aspire automatically injects environment variables into the consuming resource—this is often referred to as *configuration injection*.

These environment variable names follow specific conventions based on the referenced resource’s name and its type. Understanding these conventions is especially important for non-.NET applications (such as Python, Node.js, or Go), where you read environment variables directly instead of relying on typed Aspire client integrations.
```

var cache = builder.AddRedis("my-cache");
var db = builder.AddPostgres("postgres").AddDatabase("my-db");

var api = builder.AddProject<Projects.Api>("api")
    .WithReference(cache)
    .WithReference(db);

// After adding all resources, run the app...
builder.Build().Run();
```

The `api` resource receives the following environment variables:

| Environment variable | Description |
|---|---|
| `ConnectionStrings__my-cache` | Connection string for the Redis cache |
| `ConnectionStrings__my-db` | Connection string for the PostgreSQL database |

<Aside type="tip">

In .NET applications, you can access connection strings with `builder.Configuration.GetConnectionString("my-cache")`, which automatically resolves the `ConnectionStrings__` prefix. For non-.NET applications, read the full environment variable name directly.

</Aside>

### Endpoint URLs

When you reference a resource that exposes HTTP or HTTPS endpoints (such as a project or container service), Aspire generates environment variables for the endpoint URLs. The resource name is first **encoded** (hyphens and other non-alphanumeric characters are replaced with underscores), then **uppercased**:

> ```

Reviewer: ```suggestion
```txt
```

var cache = builder.AddRedis("my-cache");
var db = builder.AddPostgres("postgres").AddDatabase("my-db");

var api = builder.AddProject<Projects.Api>("api")
    .WithReference(cache)
    .WithReference(db);

// After adding all resources, run the app...
builder.Build().Run();
```

The `api` resource receives the following environment variables:

| Environment variable | Description |
|---|---|
| `ConnectionStrings__my-cache` | Connection string for the Redis cache |
| `ConnectionStrings__my-db` | Connection string for the PostgreSQL database |

<Aside type="tip">

In .NET applications, you can access connection strings with `builder.Configuration.GetConnectionString("my-cache")`, which automatically resolves the `ConnectionStrings__` prefix. For non-.NET applications, read the full environment variable name directly.

</Aside>

### Endpoint URLs

When you reference a resource that exposes HTTP or HTTPS endpoints (such as a project or container service), Aspire generates environment variables for the endpoint URLs. The resource name is first **encoded** (hyphens and other non-alphanumeric characters are replaced with underscores), then **uppercased**:

```txt

> {RESOURCE_NAME}_{SCHEME}

Reviewer 3: `SCHEME` is not the discriminator used here. `WithReference` uses the endpoint name (`{RESOURCENAME}_{ENDPOINTNAME}`), and it does not drop the suffix for a single endpoint. A named endpoint such as `admin` produces `MY_API_ADMIN`. Could we update the format/table and remove the single-endpoint claim?

```

If the resource only exposes a single endpoint, the scheme suffix is omitted:

```txt
{RESOURCE_NAME}
```

For example:

```csharp title="C# — AppHost.cs"
var builder = DistributedApplication.CreateBuilder(args);

var api = builder.AddProject<Projects.Api>("my-api");

var frontend = builder.AddJavaScriptApp("frontend", "./app")
    .WithReference(api);

// After adding all resources, run the app...
builder.Build().Run();
```

The `frontend` resource receives:

| Environment variable | Example value |
|---|---|
| `MY_API_HTTP` | `http://localhost:5000` |
| `MY_API_HTTPS` | `https://localhost:5001` |

### Service discovery variables

In addition to the endpoint URL variables, Aspire also generates service discovery variables for .NET service resolution. These use the format:

```txt
services__{resourcename}__{scheme}__{index}
```

> These variables use double underscores (`__`) as separators and are **lowercase**. For example, `services__my-api__http__0`. .NET applications use these automatically through the service discovery system. Non-.NET applications should use the [endpoint URL variables](#endpoint-urls) instead.

Reviewer 3: I do not think we can say these variables are lowercase. Aspire preserves the resource name as authored, so `AddProject("MyApi")` produces `services__MyApi__http__0`. Could we say the separators are fixed but the resource-name casing is preserved?

    .WithEnvironment("DB_HOST", db.Resource.Parent.PrimaryEndpoint.Property(EndpointProperty.Host))
    .WithEnvironment("DB_PORT", db.Resource.Parent.PrimaryEndpoint.Property(EndpointProperty.Port));

// After adding all resources, run the app...
builder.Build().Run();
```

This approach lets you define explicit, predictable variable names without relying on the automatic naming conventions.

<Aside type="tip">

Use the [Aspire dashboard](/dashboard/explore/#resource-details) to inspect the actual environment variables that Aspire injects into each resource. This is the fastest way to discover the exact variable names available to your application.

</Aside>

<LearnMore>

[Service discovery](/fundamentals/service-discovery/)

</LearnMore>

<LearnMore>

[Inner-loop networking overview](/fundamentals/networking-overview/)

</LearnMore>

<LearnMore>

[Python integration](/integrations/frameworks/python/)

</LearnMore>

<LearnMore>

[JavaScript integration](/integrations/frameworks/javascript/)

</LearnMore>

<LearnMore>

[Executable resources](/app-host/executable-resources/)

</LearnMore>

<LearnMore>

[Deployment manifest format](/deployment/manifest-format/)

> </LearnMore>

Reviewer: Hey @copilot, you messed this up bad. Only use LearnMore under a paragraph and never in rapid succession like that.
